VideoPress: change the logic to enqueue video block scripts

retrofox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

This PR changes the logic to enqueue the relevant scripts used by the block editor to handle the behavior of the VideoPress video block, looking for:

* Ensure do not load scripts twice or more
* Use the same handle defined by the video block to enqueue the rest of them
* Ensure loading the scripts on all platforms

It fixes a bug that happens in Simple sites. The  `videoPressEditorState` is not provided by the server side to the client, generating an error when the client tries to destructure it:

![image](https://user-images.githubusercontent.com/77539/214809149-e2c5bf5a-1179-42e5-b0b5-6a292702139d.png)

Fixes #

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* VideoPress: change the logic to enqueue video block scripts

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

Internal: p1674641635344319-slack-C01U2KGS2PQ

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to the block editor
* Ensure the `video/index` js file loads once (filter the files by `blocks/video`)
<img width="576" alt="image" src="https://user-images.githubusercontent.com/77539/214592515-261b12b8-f9e4-4ceb-8000-ca4aba25c041.png">

* Ensure the global `videoPressEditorState` var is defined
* Ensure the global `JP_CONNECTION_INITIAL_STATE` var is defined
* Test in all platforms

retrofox committed:
[not verified] move enqueue-scripts logic to block registration

retrofox committed:
[not verified] changelog

retrofox committed:
[not verified] rely on block metadata to get the handle name

sergeymitr reviewed (dismissed):
Tests well 👍

sergeymitr commented:
Approved, looks good to me.
I'm leaving the "Needs Review" label though, in case you still need a team review. Please adjust the yourself as needed.

jeherve reviewed (commented):
This gets rid of the error, and works well on self-hosted, but I do get a connection prompt on my WordPress.com simple site:

<img width="825" alt="image" src="https://user-images.githubusercontent.com/426388/214644533-16a4007a-65c0-43a2-a562-b346a83de7e7.png">

Is this expected?

dhasilva commented on projects/packages/videopress/src/class-initializer.php:
```diff
@@ -193,6 +192,11 @@ public static function register_videopress_video_block() {
+		// Register and enqueue scripts used by the VideoPress block.
+		Block_Editor_Extensions::init( $videopress_video_metadata );
+
+		// Do not register is the block is already registered.
```
```suggestion
		// Do not register if the block is already registered.
```

dhasilva commented:
> This gets rid of the error, and works well on self-hosted, but I do get a connection prompt on my WordPress.com simple site:
> 
> <img alt="image" width="825" src="https://user-images.githubusercontent.com/426388/214644533-16a4007a-65c0-43a2-a562-b346a83de7e7.png">
> 
> Is this expected?

I also see it. As an aside, I'm using the `Videomaker` theme and the message is unreadable.
![2023-01-25_16-18-24](https://user-images.githubusercontent.com/8486249/214664991-73fedbd5-dce1-46a8-a1a0-3ec69dd548ae.png)

Otherwise, the `Uncaught TypeError: window.videoPressEditorState is undefined` error is fixed and the VideoPress block is selectable.

zinigor reviewed (dismissed):
The PR looks good to me except one little thing. It would have been great if we could have the reason for the change explained in the description.

retrofox committed:
[not verified] Update projects/packages/videopress/src/class-initializer.php

Co-authored-by: Douglas Henri <dhasilva@users.noreply.github.com>

retrofox commented:
> Is this expected?

No, it isn't.

retrofox commented:
jetpack-connection doesn't behave appropriately in dotcom, setting the user is disconnected. That's why it shows the banner.

retrofox committed:
[not verified] Set connection status based also on the site type
Also, rely only in the initial blobal state to know the connectiom status for the user, avoing to use the useConnection() hook.

retrofox committed:
set banner color

retrofox commented:
> I also see it. As an aside, I'm using the Videomaker theme and the message is unreadable.

Thanks. Nice one. Addressed here https://github.com/Automattic/jetpack/pull/28564/commits/59a0df79d2abdc2901e3f50a9e599cb79de55e81

retrofox commented:
> The PR looks good to me except one little thing. It would have been great if we could have the reason for the change explained in the description.

👍 Updated the PR description. Thanks.

jeherve commented on projects/packages/videopress/src/client/block-editor/blocks/video/edit.tsx:
```diff
@@ -45,7 +44,13 @@ import './editor.scss';
 
-const { myJetpackConnectUrl } = window.videoPressEditorState;
+// Get site type.
+const { siteType, myJetpackConnectUrl } = window.videoPressEditorState;
```
Would it be worth to add some defaults, just in case?

```suggestion
const { siteType = '', myJetpackConnectUrl } = window?.videoPressEditorState || {};
```

retrofox replied:
Well, that will help avoid breaking the block in case videoPressEditorState is not defined. But we need to ensure it's defined. So in some way, we'll silence the error. 😬

retrofox replied:
commited

retrofox committed:
Update projects/packages/videopress/src/client/block-editor/blocks/video/edit.tsx

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

retrofox commented:
> I also see it. As an aside, I'm using the `Videomaker` theme and the message is unreadable. ![2023-01-25_16-18-24](https://user-images.githubusercontent.com/8486249/214664991-73fedbd5-dce1-46a8-a1a0-3ec69dd548ae.png)

<img width="861" alt="image" src="https://user-images.githubusercontent.com/77539/214815237-af4d8229-fe79-47cd-bde0-ea2402fa0dd4.png">

lhkowalski commented on projects/packages/videopress/src/client/block-editor/blocks/video/edit.tsx:
```diff
@@ -45,7 +44,13 @@ import './editor.scss';
+// Get site type.
+const { siteType = '', myJetpackConnectUrl } = window?.videoPressEditorState || {};
+
+// Get connection intial state from the global window object.
```
```suggestion
// Get connection initial state from the global window object.
```

lhkowalski committed:
Fix typo on initial state comment

lhkowalski commented:
Tested the change on a Jetpack site. The global vars are present:

<img width="700" alt="Screen Shot 2023-01-26 at 10 13 47" src="https://user-images.githubusercontent.com/6760046/214844450-299b81dd-9dcf-4473-911c-2f86e0f247cc.png">

The script is loaded once as well:

<img width="500" alt="Screen Shot 2023-01-26 at 10 09 50" src="https://user-images.githubusercontent.com/6760046/214843680-ef30ae75-203a-47fd-9627-a1a10de5a8fb.png">

lhkowalski commented:
Also tested the changes on a Simple Site.

The vars are present:

<img width="700" alt="Screen Shot 2023-01-26 at 10 11 21" src="https://user-images.githubusercontent.com/6760046/214844011-2ab74ff0-3f14-4272-b1a4-7a7b213f657f.png">

The script is also loaded only once:

<img width="500" alt="Screen Shot 2023-01-26 at 10 10 57" src="https://user-images.githubusercontent.com/6760046/214844088-2c893170-463f-4a08-acec-c659e7f91c0a.png">

lhkowalski commented:
Tested on Atomic as well.

The script is loaded once:

<img width="400" alt="Screen Shot 2023-01-26 at 10 29 51" src="https://user-images.githubusercontent.com/6760046/214847630-9e3f531a-d2a8-4071-8c46-b9c3a746c644.png">

The vars are present:

<img width="700" alt="Screen Shot 2023-01-26 at 10 26 51" src="https://user-images.githubusercontent.com/6760046/214847210-4067ebe1-96a9-492c-aee8-cb6fd61d7ab1.png">

lhkowalski reviewed (approved):
Code changes are good and the tests generated the expected results on a Jetpack site, a Simple Site and an Atomic site.
